src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java

```diff
@@ -182,61 +186,83 @@ private boolean isTerminated(final DetailAST ast, boolean useBreak,
                 terminated = true;
                 break;
             case TokenTypes.LITERAL_BREAK:
-                terminated = useBreak;
+                terminated = useBreak || hasOuterLabel(ast, labels);
                 break;
             case TokenTypes.LITERAL_CONTINUE:
-                terminated = useContinue;
+                terminated = useContinue || hasOuterLabel(ast, labels);
                 break;
             case TokenTypes.SLIST:
-                terminated = checkSlist(ast, useBreak, useContinue);
+                terminated = checkSlist(ast, useBreak, useContinue, labels);
                 break;
             case TokenTypes.LITERAL_IF:
-                terminated = checkIf(ast, useBreak, useContinue);
+                terminated = checkIf(ast, useBreak, useContinue, labels);
                 break;
             case TokenTypes.LITERAL_FOR:
             case TokenTypes.LITERAL_WHILE:
             case TokenTypes.LITERAL_DO:
-                terminated = checkLoop(ast);
+                terminated = checkLoop(ast, labels);
                 break;
             case TokenTypes.LITERAL_TRY:
-                terminated = checkTry(ast, useBreak, useContinue);
+                terminated = checkTry(ast, useBreak, useContinue, labels);
                 break;
             case TokenTypes.LITERAL_SWITCH:
-                terminated = checkSwitch(ast, useContinue);
+                terminated = checkSwitch(ast, useContinue, labels);
                 break;
             case TokenTypes.LITERAL_SYNCHRONIZED:
-                terminated = checkSynchronized(ast, useBreak, useContinue);
+                terminated = checkSynchronized(ast, useBreak, useContinue, labels);
+                break;
+            case TokenTypes.LABELED_STAT:
+                labels.add(ast.getFirstChild().getText());
+                terminated = isTerminated(ast.getLastChild(), useBreak, useContinue, labels);
                 break;
             default:
                 terminated = false;
         }
         return terminated;
     }
 
+    /**
+     * Checks if a given break statement is terminating a switch statement that is labeled.
+     *
+     * @param breakAst the break statement to check.
+     * @return true if the break statement is terminating a labeled switch statement.
+     */
+    private static boolean isLabeledBreakTerminatingSwitch(final DetailAST breakAst) {
+        final DetailAST literalBreakLabel = breakAst.findFirstToken(TokenTypes.IDENT);
+        return literalBreakLabel != null;
+    }
+
+    /**
+     * Checks if given break or continue ast has outer label.
+     *
+     * @param ast break or continue node
+     * @param labels label names in scope of case group
+     * @return true if local label used
+     */
+    private static boolean hasOuterLabel(DetailAST ast, Set<String> labels) {
+        return ast.getChildCount() == 2 && !labels.contains(ast.getFirstChild().getText());
```

Comment:
> Please extract `ast.getChildCount() == 2` to some meaningfully named variable so that the next person to code in here knows why this matters.

Specifically, on observing for the below piece fo code which leads to Fall-through false positive, the child count is 3. 
```
 case 2:
                inner: for (int i = 0;i < 3;i++) {
                    break outer2; // ok
                }
```

Some logs from the console:
```
Child 0: Type = 86, Text = break
Child 1: Type = 144, Text = //
Child 2: Type = 73, Text = }
```